### PR TITLE
rename psroi_pooling -> ps_roi_average_pooling and refactor

### PR DESCRIPTION
**Pull request by knorth55**
split from #773 

- rename `psroi_pooling` -> `ps_roi_average_pooling_2d`

### CODE REVIEW

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 1]
```diff
@@ -74,47 +79,61 @@ def forward_cpu(self, inputs):
...
+        pooled_dim = self.out_c
+        pooled_height = self.out_h
+        pooled_width = self.out_w
+        group_size = self.group_size
+
+        for i in six.moves.range(top_data.size):
+            pw = i % pooled_width
+            ph = int(i / pooled_width) % pooled_height
+            ctop = int(i / pooled_width / pooled_height) % pooled_dim
+            n = int(i / pooled_width / pooled_height / pooled_dim)
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]: How about this?
```python
n, ctop, ph, pw = np.unravel_index(i, top_data.shape)
```

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 2]
```diff
@@ -207,46 +222,57 @@ def forward_gpu(self, inputs):
...
+        pooled_dim = self.out_c
+        pooled_height = self.out_h
+        pooled_width = self.out_w
+        group_size = self.group_size
+
+        for i in six.moves.range(top_diff.size):
+            pw = i % pooled_width
+            ph = int(i / pooled_width) % pooled_height
+            ctop = int(i / pooled_width / pooled_height) % pooled_dim
+            n = int(i / pooled_width / pooled_height / pooled_dim)
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]: How about this?
```
n, ctop, ph, pw = np.unravel_index(i, top_data.shape)
```

Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

File: chainercv/functions/ps_roi_average_pooling_2d.py (hunk 1 above)
Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 3]
```diff
@@ -207,46 +222,57 @@ def forward_gpu(self, inputs):
...
+            c = (ctop * group_size + gh) * group_size + gw
+
+            if (hstart >= hend) or (wstart >= wend):
+                continue
+
+            count = (hend - hstart) * (wend - wstart)
+            diff_val = top_diff[n, ctop, ph, pw] / count
+            for y in six.moves.range(hstart, hend):
+                for x in six.moves.range(wstart, wend):
+                    bottom_diff[roi_batch_ind, c, y, x] += diff_val
```
[the remark this answers is not in the document]
Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 4]
```diff
@@ -74,47 +79,61 @@ def forward_cpu(self, inputs):
...
+                top_data[n, ctop, ph, pw] = 0
+                continue
+
+            output_val = 0.
+            count = (hend - hstart) * (wend - wstart)
+            for y in six.moves.range(hstart, hend):
+                for x in six.moves.range(wstart, wend):
+                    output_val += bottom_data[roi_batch_ind, c, y, x]
+            output_val /= count
+            top_data[n, ctop, ph, pw] = output_val
```
[the remark this answers is not in the document]
Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

**Comment by Hakuyume** [comment]
@knorth55 Could you merge `master` into this PR to enable pfnCI?

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 5]
```diff
@@ -74,47 +79,58 @@ def forward_cpu(self, inputs):
...
+            if hstart >= hend or wstart >= wend:
+                top_data[n, ctop, ph, pw] = 0
+                continue
+
+            output_val = 0.
+            count = (hend - hstart) * (wend - wstart)
+            for y in six.moves.range(hstart, hend):
+                for x in six.moves.range(wstart, wend):
+                    output_val += bottom_data[roi_batch_ind, c, y, x]
+            output_val /= count
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]: The computation of `output_val` is no longer needed.

Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

**Comment by knorth55** [comment]
I merged `origin/master`

**Comment by Hakuyume** [comment]
pfnCI, test this please

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+        bottom_data, bottom_rois, bottom_roi_indices = inputs
+        channel, height, width = bottom_data.shape[1:]
+        if self.out_c is None:
+            if channel % (self.group_size * self.group_size) != 0:
+                raise ValueError(
+                    'input channel must be divided by group_size * group_size:'
+                    '{} % {} != 0'
+                    .format(channel, self.group_size * self.group_size))
+            out_c = int(channel / (self.group_size * self.group_size))
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]:
```python
out_c = channel // (self.group_size * self.group_size)
```

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+import numpy as np
+import six
+
+import chainer
+from chainer.backends import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _pair(x):
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]: The name of this helper should be improved.

File: chainercv/functions/ps_roi_average_pooling_2d.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+
+        bottom_data, bottom_rois, bottom_roi_indices = inputs
+        channel, height, width = bottom_data.shape[1:]
+        if self.out_c is None:
+            if channel % (self.group_size * self.group_size) != 0:
+                raise ValueError(
+                    'input channel must be divided by group_size * group_size:'
+                    '{} % {} != 0'
+                    .format(channel, self.group_size * self.group_size))
+            out_c = int(channel / (self.group_size * self.group_size))
```
Comment (Hakuyume) [inline chainercv/functions/ps_roi_average_pooling_2d.py]: ditto.

Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

File: chainercv/functions/ps_roi_average_pooling_2d.py (hunk 6 above)
Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: updated

File: chainercv/functions/ps_roi_average_pooling_2d.py (hunk 7 above)
Comment (knorth55) [reply chainercv/functions/ps_roi_average_pooling_2d.py]: rename to `_outsize`

**Comment by knorth55** [comment]
@Hakuyume can you review this again?

**Comment by Hakuyume** [comment]
pfnCI, test this please

**Comment by Hakuyume** [comment]
pfnCI, test this please

**Pull request merged**